### ISSUE TITLE
Channel interface needs shutdown/close

### ISSUE THREAD

**Issue by louiscryan** [grpc/grpc-java#64]
With the elimination of Service the Channel interface is now insufficient for normal use. 

While ChannelImpl has a shutdown Channel does not and intercepting a ChannelImpl immediately converts it into Channel.

Closeable/AutoCloseable would be fine too

**Comment by ejona86** [comment]
This issue keeps coming up, and I have discussed offline with people my concerns with `shutdown()` on `Channel`. Writing a bit of it here for future sharing.

My main concern with `shutdown()` on Channel is that is breaks knowing what is `shutdown()`. With interceptors and stubs, you form "trees" of Channels with leaves being Channels/stubs used directly and the root being a ChannelImpl. Because of the many `Channel`s that all use the same `ChannelImpl`, you can't predict if other Channels will break if you `shutdown()` another `Channel`.

To make it more clear, an example:
```java
ChannelImpl channelImpl = NettyChannelBuilder.forAddress("example.com", 443).build();
Channel channel1 = ClientInterceptors.intercept(channelImpl,
    new ClientAuthInterceptor(creds1, executor));
Channel channel2 = ClientInterceptors.intercept(channelImpl,
    new ClientAuthInterceptor(creds2, executor));
channel2.shutdown();
// Now channel1 is also shutdown
```

Note that that is assuming `ClientInterceptor`s can't implement `shutdown()`. If they _can_ (or if someone made an interceptor implementing `Channel` directly), then the problem becomes more acute as now the interceptor for `channel1` could be expecting `shutdown()` to be called. Such a design would require all channels to be `shutdown()`, even though all `Channel`s with the same root would be "shutdown" as soon as the first one was. This leads to indeterminate states where the `ClientInterceptor` hasn't been shutdown but the underlying `ChannelImpl` has.

Channels are expected to be injected (manual, or automatic) in many cases. When injected, you ideally wouldn't want them to have `shutdown()` because the code receiving the object doesn't know if anyone else is using it.

Basically, only the creator of the _`ChannelImpl`_ knows the its appropriate lifecycle. Although it would be _convenient_ in simple examples to not have to store a reference to the `ChannelImpl`, it would only work well when you know there is only one interceptor chain. However, it is explicitly part of the interceptor design to permit the "trees" to form. Permitting and encouraging users to `shutdown()` a `Channel` would seem to be a disservice to them.

**Comment by wrwg** [comment]
I think some of the confusion comes from the naming. A class called FooImpl indicates for me to have its main purpose in implementing interface Foo. That leads me to think I can just work with the Foo.

A better naming may make the design clearer. Channel is the thing on which you can initiate a call and which allows interceptors to plug in. ChannelImpl might be better called TransportChannel, ConnectedChannel, or something which indicates its a resource. That may make it more obvious you have to take care of it for clean up.

**Comment by zhangkun83** [comment]
@wrwg I feel the same. +1 for renaming ChannelImpl.

**Comment by ejona86** [comment]
I'm game for a rename to. The only reason it's been ChannelImpl is lack of a better name. TransportChannel and ConnectedChannel are both worse than ChannelImpl, in my mind. DefaultChannel seems little better than ChannelImpl.

**Comment by zhangkun83** [comment]
IMO DefaultChannel is as bad as ChannelImpl. "Default" is not conveying anything useful (default in what sense?), other than indicating it's the "default impl". Since ClientTransport is "encapsulating a single connection" (quoted from its javadoc), and ChannelImpl is based on a ClientTransport,  "TransportChannel" and "ConnectionChannel" are more informative and less misleading.

**Comment by nmittler** [comment]
I think the problem here is that the term "channel" ~= "transport", but an "interceptor" != "transport". I think it would be better to provide a more abstract interface that is responsible for creating calls and then all of these "things" would implement/extend that. Something like this:

```java
interface CallFactory {
   Call newCall(MethodDescriptor method);
}
```

```java
interface Channel extends CallFactory {
    ...
    void shutdown();
}
```

```java
class ChannelImpl implements Channel {
    ...
}
```

**Comment by ejona86** [comment]
ConnectionChannel and TransportChannel are worse than DefaultChannel and ChannelImpl in my mind.  Connection, Transport, and Channel in many conversations (obviously not ours) are interchangeable. Connection and Transport don't add any additional meaning to Channel, yet it isn't obviously filler. When you see Default\* or \*Impl you can quickly ignore the filler.

We only have one Channel implementation. If we ever have more than one, then maybe this problem would be solved with SimpleChannel, LoadBalancedChannel, etc. However, I still see the value of only having one implementation of Channel. If we stay with one, maybe we rename Channel to CallFactory and rename ChannelImpl to Channel.

**Comment by nmittler** [comment]
@ejona86 I don't see the harm in having an interface for `Channel` even though there is only (currently) a single instance. The broader point that I was trying to make is that thinking of `interceptors` as kinds of "channels" is wrong, since they do not represent the transport ... they only represent the ability to create new calls (i.e. on an existing transport).

**Comment by ejona86** [comment]
@nmittler, the harm is that there is no distinction between Channel the interface and Channel the implementation. If there is only one implementation with no additional methods compared to the interface, then why bother having the interface? Even in tests we would want users to use the real Channel implementation.

We have trouble coming up with a name because they represent the same identical idea. If there were multiple implementations, then the name becomes whatever is special about that implementation. With only one implementation, then you get something like DefaultChannel or ChannelImpl. As seen in my previous comment, if there are multiple implementations naming them is _easy_.

**Comment by wrwg** [comment]
I agree with @nmittler's direction. If we consider channel and transport as too easily mixed up notions for users, perhaps the best call is to rename the interface Channel into something more neutral, like `CallFactory`.

In the process we could get rid of the `Channel` interface and rename `ChannelImpl` to `Channel`.

**Comment by nmittler** [comment]
@ejona86 The renaming that I suggested won't solve this issue, agreed ... but it goes some length at clarifying these interfaces a bit.

So we still have the problem that creating the interceptor chain will yield a `CallFactory` now, rather than a `Channel`.  Perhaps the way to solve this is by defining another class representing the interceptor chain:

```java
public class InterceptorChain implements CallFactory {
    ...
    public Channel getChannel();
}
```

**Comment by zhangkun83** [comment]
I think we need a type name that can be passed along and used to make RPC calls. Channel sounds a fine name. CallFactory describes the capability perfectly, but it sounds weird as a replacement of Channel and something with similar functionality as Stub.

The Channel interface is also the center of the "Channel Layer API" that we talk about a lot. Moving away from it requires unnecessary mind-shift.

"SimpleChannel" sounds a fine new name for ChannelImpl, and it matches well with "LoadBalancingChannel".

**Comment by nmittler** [comment]
@zhangkun83 I disagree ... I don't think the mind-shift is unnecessary or all that out of place in the "channel API". The overloaded term "Channel" has been causing many of us problems WRT interceptors ... since they're not channels except that they happen to implement the interface. Changing the "thing" that our stubs use to `CallFactory` represents, in my mind, a clarification of the requirements of the stub ... nothing more.

**Comment by ejona86** [comment]
@nmittler, that does solve most of the problems. But it does turn a light-weight interface into a heavy-weight interface. Also, it also wouldn't be as good for injection, because you don't want the injected code to use getChannel. Note also that the Channel implementation would have to implement InterceptorChain, which is icky (but mainly a naming issue).

**Comment by nmittler** [comment]
@ejona86 I"m not sure I understand the points your'e making.  Why would the `Channel` have to implement `InterceptorChain`?

**Comment by zhangkun83** [comment]
@nmittler Just my gut feeling ... when you have an abstraction layer, you need an umbrella term for it. That term should not be too specific, because what it includes may change in the future. Transport, Channel, and Stub are all non-specific terms, and they will tolerate changes well. CallFactory is too specific for this purpose.

**Comment by nmittler** [comment]
Maybe the name of the "layer" is the problem.  Really, we should be referring to it as the "Call" layer, not "Channel".

**Comment by nmittler** [comment]
Thinking about this some more, I question the value of `Channel`'s is-a relationship with `CallFactory`.  I think composition might be more appropriate:

```java
public class Channel {
    // Create the channel with optional interceptors that are "global" to all calls for
    // this channel.
    Channel(TransportFactory tf, ExecutorService es, Interceptor... interceptors); 

    // The call factory used for this channel. If interceptors were supplied, they
    // will be applied to any calls created by this factory.
    CallFactory callFactory();

    void shutdown();
    ...
}
```

I believe this would also address #538.  Thoughts?

**Comment by ejona86** [comment]
@nmittler, if `InterceptorChain` can return the base `Channel`, then `Channel` must be an `InterceptorChain`. Otherwise, you have no common type between `Channel` and `InterceptorChain` that you can pass to other code _and_ still allow that code to add interceptors on top. Really, if we are going to have `getChannel()` I'd expect it would need to be on `CallFactory`, otherwise `CallFactory` is useless, as you couldn't add interceptors (and thus couldn't use it for stubs).

We could get rid of the is-a relationship between `Channel` and `CallFactory`, but I don't think it gains us much. I don't think it helps #538, since I don't see any reason the is-a relationship would prevent the changes needed for #538.

**Comment by nmittler** [comment]
@ejona86 I'm don't quite follow your reasoning ... I suspect that I don't understand the problem you're trying to solve.

`Channel` as the thing that controls connection state. The connection lifecycle is orthogonal to the interceptor lifecycle. We may very well have interceptor chains that will be "shutdown/closed", but that would have no affect on the underlying `Channel`. This would need to be the case since you could have several interceptor chains that use the same underlying `Channel` (e.g. one chain per stub). The application will ultimately have to shutdown the channel as a separate step, after it has completed all calls across all stubs.

**Comment by ejona86** [comment]
@nmittler, shutting down interceptor chains does not work, unless you want to start doing reference counting and forcing every stub to require explicit closure. I am actually confused what you are responding to, since your `CallFactory` and `InterceptorChain` both don't have a shutdown method.

**Comment by nmittler** [comment]
@ejona86 I wasn't actually suggesting that we provide lifecycle handling for interceptors. I was referring to the "interceptor lifecycle" only to make the point that they simply can't have control over the channel lifecycle.  That is the responsibility of the application. The "thing" returned by creation of an interceptor chain should not be a `Channel` if we want to allow multiple chains to use the same underlying channel (which I think we all agree, we do).

**Comment by ejona86** [comment]
@nmittler, if by `Channel` you meant a concrete class `Channel` (`ChannelImpl` today), then I agree with no reservations.

My point was if you are going to pass an object to some other class, what type do you pass? If you pass a `CallFactory`, then the class can't add interceptors. If you pass `InterceptorChain` then you can't pass `Channel` (`ChannelImpl`) without any interceptors. And passing `Channel` (`ChannelImpl`) doesn't let you add interceptors.

**Comment by nmittler** [comment]
@ejona86 and I discussed a bit offline.  IIUC, there are essentially 2 use cases for interceptors:

1. Allowing global interceptors for the `Channel`. These can be installed as part of the process of building the `Channel`. These interceptors would be used for all stubs that use the channel. I believe this solves this particular issue for single-stub scenarios. If the channel has built-in interceptors, the "thing" that's returned is a `Channel` with a `shutdown` method.
2. If the stubs chose to decorate the channel with their own interceptors, those would be added to the head of the interceptor chain such that the execution would be stub interceptors, followed by channel interceptors. In this case, the application will have to maintain a reference to the `Channel` that is shared by the stubs and shut it down manually when all calls have completed.

With that said, I do think that it's worthwhile to address the renaming. We can tackle that as a separate issue.

WDYT?

**Comment by ejona86** [comment]
@nmittler, the approach you described is #538. If that approach is what we are going to go with, then this issue could be closed.

I'm okay with the renaming, but I do see it as being separate, as it doesn't add any previously-unavailable functionality. Both this issue and #538 are for functional deficiencies. I created #545 for tracking the renaming.

**Comment by nmittler** [comment]
@louiscryan since you opened this issue, what are your thoughts? :)

**Comment by wrwg** [comment]
Making one observation @nmittler brought up first, I think, explicit.

The interface:

```
interface Channel {
  Call newCall(MethodDescriptor);
}
```

... is in his it's sole existence odd. Looking at  this standalone, there is nothing in here which makes this a "channel". The interface is not named for what it actually does, but for what is its purpose in some scenarios (and in others like interceptor chaining the name would not even fit that well).

**Comment by ejona86** [comment]
@wrwg, for discussions regarding naming, please use #545 now that it exists. Let's leave this issue and #538 for the functional problems with the current API.

**Comment by ejona86** [comment]
Mostly addressed by adding interceptors into builder.

**Comment by ramaraochavali** [comment]
@ejona86 I know this issue has been closed long back. But just want to get your opinion on adding `getChannel()` method on `InterceptorChannel` to get access to the original `ManagedChannel` so that applications can call `shutdown` on it. Right now we are build stub with lots of interceptors etc, at one point we only have access to stub object - and we are not able to call shutdown on it because the stub's `getChannel` returns `InterceptorChannel`. If we have a method on `InterceptorChannel` to get the original channel, we could get traverse the tree till we get the `ManagedChannel`. What do you think?

**Comment by ejona86** [comment]
@ramaraochavali, I'd suggest you to save a reference to the `ManagedChannel`. A key detail of a stub is you don't need to shut it down. If you add the ability to shut it down then it becomes confusing on whether you are _expected_ to shut it down, and then you have different pieces of code shutting it down or shutting it down prematurely for some other code.

**Comment by ramaraochavali** [comment]
@ejona86  I am not suggesting to add `shutdown` on Stub. Sorry for mis-communication above. What I am suggesting is to add the above methods in `InterceptorChannel` so that we could use the following 
Channel channel = stub.getChannel()
ManagedChannel originalChannel = null;
if (channel instanceof ManagedChannel) {
   originalChannel = (ManagedChannel) channel;
} else if (channel instanceof InterceptorChannel) {
  originalChannel = ((InterceptorChannel) channel). getParentManagedChannel(); //We could find a better name for this method
}
originalChannel.shutdown();

**Comment by ejona86** [comment]
> I am not suggesting to add shutdown on Stub.

I didn't think you were. The point is that some stubs will need their managed channel shutdown and others won't. That gets confusing, and it'll be unclear which you should do in many situations. Is there a reason you can't pass the ManagedChannel to the code that needs to shut it down?

We had a discussion similar to this around the time of 0.15, and decided it was confusing and gained little.

**Comment by ramaraochavali** [comment]
@ejona86 We have actually abstracted the `Channel` from consumers to simplify the usage. So we create `ManagedChannel` based on destination taken from user in our custom `ClientBuilder`. So the consumers do not have access to `Channel` - which we are thinking in parallel to see if it would be good to expose or not.
We are also considering your suggestion of keeping a reference of `ManagedChannel`. The way we solved today is by introducing an extension of `ManagedChannel` that holds the reference of the original `ManagedChannel` and tip of the interceptor chain i.e. `InterceptorChannel`. This class delegates the `Channel` related calls to interceptor tip and `ManagedChannel` related calls to the original channel. We pass this extended class to stub during building of Stub. 

But the point I am trying to make is , since `Stub` already has a reference to the original managed channel, we could provide a getter in `Stub` (which I prefer) or in `InterceptorChannel` and let people get it shutdown the channel whenever they want. I understand your point about `Stub` and `Channel` separation though as multiple stubs can share the same `Channel` depending on how it is setup. But I see no harm in exposing the following api  in `AbstractStub` - It is user's choice on how to use it.

```
ManagedChannel getManagedChannel() {
     Channel channel = this.getChannel();
    ManagedChannel managedChannel = null;
    if (channel instanceof ManagedChannel) {
        managedChannel = (ManagedChannel) channel;
    } else if (channel instanceof InterceptorChannel) {
             //Iterate the InceptorChain till we get the ManagedChannel.
              while(channel instanceof InterceptorChannel) {
                 channel = ((InterceptorChannel) channel).getChannel();
              }
             if (channel instanceof ManagedChannel) {
                 managedChannel = (ManagedChannel) channel;
            }
    }
    return managedChannel;
}
```
Do you see any problem with this?

**Comment by ramaraochavali** [comment]
@ejona86 just following-up on this? What do you think about this?

**Comment by ejona86** [comment]
> But I see no harm in exposing the following api in AbstractStub - It is user's choice on how to use it. 

The "user's choice" is the problem, because there's actually multiple users and they have to coordinate. There is the user creating the ManagedChannel and the user using the Stub. With this added "choice" a receiver of a Stub (which may not even be the end-consumer! It could be a utility that modifies the stub such that now there are two!) has a new question "do I shut this down?". That's the source of the complexity and confusion.

It's a bit unclear what you're saying you're doing today. But it sounds like wrapping the intercepted channel with something like this before passing it to the stub isn't too bad. Although I would agree it has the limitation that user must keep the original stub (before adding any additional interceptors):
```java
public final class AccessToManagedChannelChannel extends Channel {
  private final Channel delegate;
  private final ManagedChannel managedChannel;

  public AccessToManagedChannelChannel(Channel delegate, ManagedChannel managedChannel) {
    this.delegate = delegate;
    this.managedChannel = managedChannel;
  }

  @Override public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(
      MethodDescriptor<ReqT,RespT> methodDescriptor, CallOptions callOptions) {
    return delegate.newCall(methodDescriptor, callOptions);
  }

  @Override public String authority() {
    return delegate.authority();
  }

  public getManagedChannel() {
    return managedChannel;
  }
}
```

**Comment by ramaraochavali** [comment]
@ejona86 we did some thing like this

```
public class ManagedInterceptorChannel extends ManagedChannel {

  private ManagedChannel managedChannel;
  private Channel interceptorDelegate;

  public ManagedInterceptorChannel(ManagedChannel managedChannel, Channel interceptorDelegate) {
    Preconditions.checkNotNull(managedChannel, "managedChannel");
    this.managedChannel = managedChannel;
    this.interceptorDelegate = interceptorDelegate;
  }

  @Override
  public ManagedChannel shutdown() {
    return managedChannel.shutdown();
  }

  @Override
  public boolean isShutdown() {
    return managedChannel.isShutdown();
  }

  @Override
  public boolean isTerminated() {
    return managedChannel.isTerminated();
  }

  @Override
  public ManagedChannel shutdownNow() {
    return managedChannel.shutdownNow();
  }

  @Override
  public boolean awaitTermination(long timeout, TimeUnit timeUnit) throws InterruptedException {
    return managedChannel.awaitTermination(timeout,timeUnit);
  }

  @Override
  public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
      MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
    //If there are no interceptors, directly pass this through ManagedChannel.
    if (interceptorDelegate != null) {
      return interceptorDelegate.newCall(methodDescriptor, callOptions);
    } else {
      return managedChannel.newCall(methodDescriptor, callOptions);
    }
  }

  @Override
  public String authority() {
    //If there are no interceptors, directly pass this through ManagedChannel.
    if (interceptorDelegate != null) {
      return interceptorDelegate.authority();
    } else {
      return managedChannel.authority();
    }
  }

  /**
   * Returns the underlying ManagedChannel.
   * @return ManagedChannel
   */
  public ManagedChannel getManagedChannel() {
    return managedChannel;
  }
}
```
The way people use it is like this
```
final ManagedChannelBuilder channelBuilder = channelBuilder(destination());
final ManagedChannel managedChannel = channelBuilder.build();
//Add interceptors here...
final Channel channel = ClientInterceptors.interceptForward(managedChannel, buildStandardInterceptors());
ManagedInterceptorChannel managedInterceptorChannel = new ManagedInterceptorChannel(managedChannel,channel);
Stub blockingStub = SampleGrpc.newBlockingStub(managedInterceptorChannel);
```
Some thing similar to what you have proposed? If you think it would be useful addition to grpc, I am happy to contribute this class.
